### DISCUSSION TITLE
Force element to draw on top of peers?

### DISCUSSION THREAD

**Discussion by rscarson** [slint-ui/slint#6303] [General]
I am working on the following, a hover-hint element I can attach to things
Should I give up on this approach, since it draws beneath anything set after it and instead have a single element that passes callbacks around? Or is there a way to force it to draw over other elements

```slint
export component HoverHint inherits TouchArea {
    property <bool> should_show: root.text != "" && root.has-hover;

    in-out property <int> delay: 500;
    in-out property <string> text;

    Rectangle {
        height: t.height + 5px;
        width: t.width + 25px;
        x: root.mouse-x + 15px;
        y: root.mouse-y;
        border-width: 1px;
        border-radius: 2px;

        t := Text {
            text: root.text;
            font-weight: 500;
            
            states [
                hidden when !root.should_show : {
                    color: transparent;
                }

                visible when root.should_show : {
                    color: black;
                    in {
                        animate color { duration: 0; delay: 500ms; }
                    }
                }
            ]
        }
            
        states [
            hidden when !root.should_show : {
                background: transparent;
                border-color: transparent;
            }

            visible when root.should_show : {
                background: beige;
                border-color: self.background.darker(20%);
                in {
                    animate background, border-color { duration: 0; delay: 500ms; }
                }
            }
        ]
    }
}
```

**Comment by rscarson** [comment]
I have had no luck with the `z` property:
```
MyButton {
  HoverHint {
    // z here has no effect, it will always draw beneath MyButton2
  }
}
MyButton2 {
  HoverHint {
  }
}

**Comment by rscarson** [comment]
Decided to just render the label to the left instead of the right
derp

**Comment by hunger** [comment]
My naive approach would be something along these lines:

```slint
export component HoverHint inherits TouchArea {
    property <bool> should_show: root.text != "" && root.has-hover;

    in-out property <int> delay: 500;
    in-out property <string> text;

    @children

    Rectangle {
        height: t.height + 5px;
        width: t.width + 25px;
        x: root.mouse-x + 15px;
        y: root.mouse-y;
        border-width: 1px;
        border-radius: 2px;

        t := Text {
            text: root.text;
            font-weight: 500;
            
            states [
                hidden when !root.should_show : {
                    color: transparent;
                }

                visible when root.should_show : {
                    color: black;
                    in {
                        animate color { duration: 0; delay: 500ms; }
                    }
                }
            ]
        }
            
        states [
            hidden when !root.should_show : {
                background: transparent;
                border-color: transparent;
            }

            visible when root.should_show : {
                background: beige;
                border-color: self.background.darker(20%);
                in {
                    animate background, border-color { duration: 0; delay: 500ms; }
                }
            }
        ]
    }
}
```

and then use it like this:

```slint
HoverHint {
    MyButton { ... }
}
```

I hope this actually works, I admit I did not test it :-)

**Reply by bjorn** [reply]
Moving the `HoverHint` up the hierarchy is certainly a way to tackle the rendering order issue in some cases. Unfortunately it often doesn't work, since the relevant controls are usually in a hierarchy of their own. It gets even more tricky, when there are for loop and/or list views involved.

Hence, I think we need to use the `PopupWindow`, but it has other limitations. Unfortunately the addition of the `changed` handlers are still not enough to make this thing appear and disappear based on a change in hover state. Because, as soon as the popup is shown, it eats the mouse input and `has-hover` goes to `false`, so the following causes a flickering popup:
```slint
    touch-area := TouchArea {
        width: 100%;
        height: 100%;

        changed has-hover => {
            if (self.has-hover) {
                popup.show();
            } else {
                popup.close();
            }
        }
    }

    popup := PopupWindow {
        x: touch-area.mouse-x + 10px;
        y: touch-area.mouse-y + 10px;
        
        width: 100px;
        height: 20px;
        Rectangle { background: red; }
        Text { text: "Hello World"; }
    }
```
It would need to have an option where it lets mouse events go through, at least those outside of its geometry.

**Reply by bjorn** [reply]
See [demo on Slintpad](https://slintpad.com/?snippet=%2F%2F+Copyright+%C2%A9+SixtyFPS+GmbH+%3Cinfo%40slint.dev%3E%0A%2F%2F+SPDX-License-Identifier%3A+MIT%0A%0Acomponent+Delegate+inherits+Rectangle+%7B%0A++++width%3A+100%25%3B%0A++++height%3A+50px%3B%0A%0A++++background%3A+touch-area.has-hover+%3F+green+%3A+darkgreen%3B%0A%0A++++in-out+property+%3Cint%3E+index%3B%0A%0A++++touch-area+%3A%3D+TouchArea+%7B%0A++++++++width%3A+100%25%3B%0A++++++++height%3A+100%25%3B%0A%0A++++++++changed+has-hover+%3D%3E+%7B%0A++++++++++++if+%28self.has-hover%29+%7B%0A++++++++++++++++popup.show%28%29%3B%0A++++++++++++%7D+else+%7B%0A++++++++++++++++popup.close%28%29%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++%7D%0A%0A++++Text+%7B%0A++++++++text%3A+root.index%3B%0A++++%7D%0A%0A++++popup+%3A%3D+PopupWindow+%7B%0A++++++++x%3A+touch-area.mouse-x+%2B+10px%3B%0A++++++++y%3A+touch-area.mouse-y+%2B+10px%3B%0A++++++++%0A++++++++width%3A+100px%3B%0A++++++++height%3A+25px%3B%0A++++++++Rectangle+%7B+background%3A+red%3B+%7D%0A++++++++Text+%7B+text%3A+%22tooltip+%22+%2B+root.index%3B+%7D%0A++++%7D%0A%7D%0A%0Aexport+component+MainWindow+inherits+Window+%7B%0A++++width%3A+200px%3B%0A++++height%3A+200px%3B%0A++++title%3A+%22Popup+Example%22%3B%0A%0A++++VerticalLayout+%7B%0A++++++++spacing%3A+2px%3B%0A++++++++for+i+in+3%3A+Delegate+%7B%0A++++++++++++index%3A+i%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native).

**Reply by rscarson** [reply]
I think `PopupWindow` should probably have an `enabled` property to control mouse-input passthrough

**Reply by rscarson** [reply]
I have opened a related issue, #6335
